### DISCUSSION TITLE
Strange behaviour when the orm primary key is not a primary key in the database and is null

### DISCUSSION THREAD

**Discussion by Maksych** [sqlalchemy/sqlalchemy#12089] [Usage Questions]
Folowing code prints None instead of error:

```bash
[None]
```

```python
import contextlib
from collections.abc import Iterator
from uuid import UUID, uuid4

import sqlalchemy as sa
from sqlalchemy import orm as sa_orm

ENGINE = sa.create_engine("postgresql://postgres:postgres@127.0.0.1:5432/postgres")
SESSION = sa_orm.sessionmaker(ENGINE)


class BaseV1(sa_orm.DeclarativeBase):
    pass


class UserV1(BaseV1):
    __tablename__ = "user"

    id: sa_orm.Mapped[int] = sa_orm.mapped_column(primary_key=True, autoincrement=True)
    uid: sa_orm.Mapped[UUID | None] = sa_orm.mapped_column(default=None, nullable=True)


class BaseV2(sa_orm.DeclarativeBase):
    pass


class UserV2(BaseV2):
    __tablename__ = "user"

    uid: sa_orm.Mapped[UUID] = sa_orm.mapped_column(default=uuid4, primary_key=True)


@contextlib.contextmanager
def user_v1() -> Iterator[None]:
    try:
        BaseV1.metadata.create_all(ENGINE)

        with SESSION() as session:
            session.add(UserV1())
            session.commit()

        yield
    finally:
        BaseV1.metadata.drop_all(ENGINE)


with user_v1(), SESSION() as session:
    print(session.scalars(sa.select(UserV2)).all())

```

**Comment by zzzeek** [comment]
that would be due to this logic here:

https://github.com/sqlalchemy/sqlalchemy/blob/fa76fbda4aa2887d001e1e15f9bfe9dfb24fadc8/lib/sqlalchemy/orm/loading.py#L1109C1-L1112C32

the None returned intends for this load to be a secondary object identity in the row which would be from a joined eager load.   It's nonsensical for this codepath to occur from a non-OUTER JOIN row since primary keys cannot be fully NULL.  but since you seem to be intentionally confusing the ORM with a nonsensical mapping, you're getting back the None here.

**Reply by Maksych** [reply]
Ok, if we say that's correct, does sqlalchemy have a function to check the consistency of the orm model with the db table?

**Reply by zzzeek** [reply]
it wont allow a flush (INSERT, UPDATE) of an object that has a null primary key, so it checks in that direction.    as for rows that come from the DB, it assumes primary key attributes wont be NULL.

the case you have here would be nicer to raise, if we detected that we're not in a joinedload, but it's just an unusual codepath to discover here.

**Reply by Maksych** [reply]
Yes, it's an unusual way of coding, but we don't live in a perfect world. Sometimes migrations can get interrupted and something like this can happen. In my case, a migration that deletes the primary id key and updates all uids, setting uid as the primary key, was interrupted and I got this behaviour. A raise would be great here, because of error None type object does not have an x attribute. The type checker does not know that it can be None.
